pandas/core/window.py

```diff
@@ -354,6 +354,116 @@ def _center_window(self, result, window):
                 result = np.copy(result[tuple(lead_indexer)])
         return result
 
+    def _apply(
+        self, func, name=None, window=None, center=None, check_minp=None, **kwargs
+    ):
+        """
+        Rolling statistical measure using supplied function.
+
+        Designed to be used with passed-in Cython array-based functions.
+
+        Parameters
+        ----------
+        func : str/callable to apply
+        name : str, optional
+           name of this function
+        window : int/array, default to _get_window()
+        center : bool, default to self.center
+        check_minp : function, default to _use_window
+
+        Returns
+        -------
+        y : type of input
+        """
+        if center is None:
+            center = self.center
+
+        if check_minp is None:
+            check_minp = _use_window
+
+        blocks, obj, index = self._create_blocks()
+        block_list = list(blocks)
+
+        is_window = isinstance(self, Window)
```

Comment:
Not sure if we have alternative to isinstance check

Reply:
better would be to make a method that can just be called here that can be defined in Window and Rolling and will just return all the things you need, something like

window, index, indexi = self._get_window_indexer(...)

we might also want to then rename/refactor things to be as common as possible (_prep_window and so on)

Reply:
i see u r also calling below based on is_window

may need to again have a method that u override

Reply:
This would be more modular if indexer and the kernel could be supplied independently. See for example https://github.com/pandas-dev/pandas/issues/26959, where `rolling` doesn't provide a suitable indexer, and since the indexer generation is hardcoded in, the user can't supply their own and benefit from the rest of the rolling machinery, like efficient add/remove calculations and the canned apply loop.

I was planning to do this myself, before seeing this PR.

Reply:
of course but this is the way it is now

note that others are already working on this but if it’s something reasonable would consider

Reply:
Great, Is there a PR for this you mean?  If I might do it in a followup.

Reply:
got rid of this. will add docs if changes are ok